[RFC] fix: flaky 'jobstop() kills entire process tree #6530'

erw7 opened the pull request:
Debug flaky 'jobstop() kills entire process tree #6530'.

fix https://github.com/neovim/neovim/issues/10762

blueyed commented on ci/build.ps1:
```diff
@@ -115,24 +115,25 @@ cmake --build . --config $cmakeBuildType -- $cmakeGeneratorArgs ; exitIfFailed
 
 # Functional tests
 # The $LastExitCode from MSBuild can't be trusted
+$env:TEST_FILE = "test/functional/core/job_spec.lua"
```
You can also specify `TEST_FILTER` (to not having to comment other tests).
Also `BUSTED_ARGS=--repeat=100` might be good.

blueyed commented:
Hmm.. pushed 2 commits here, but they do not show up?!

blueyed commented:
@erw7 
Triggered after 65 attempts: https://ci.appveyor.com/project/neovim/neovim/builds/26727966/job/3jf5ttm0ulw5x971?fullLog=true#L8040

blueyed commented:
`vctip.exe` showing up causes this: https://ci.appveyor.com/project/neovim/neovim/builds/26727966/job/3jf5ttm0ulw5x971?fullLog=true#L7648

blueyed commented:
I would not expect it to be a child process, is it?

Or is `os_proc_children` not working correctly?  https://github.com/erw7/neovim/blob/27565ac6f4944e7ad705243453b7d66aaad4405e/src/nvim/os/process.c#L120-L140

blueyed commented:
Also `conhost.exe` being included as a child proc already (causing the different assertion in the first place) seems incorrect to me.

justinmk commented:
Nice catch! `vctip.exe` is "Microsoft VC compiler and tools experience improvement data uploader". Sounds like a random clippy-thing.

> I would not expect it to be a child process, is it?
> 
> Or is `os_proc_children` not working correctly?

`tasklist` shows it for that ppid (parent PID), right? 

If Windows/MSVC is starting random child processes, I think it's fine to just allow 0-1 extra children in the test case. 

> Also `conhost.exe` being included as a child proc already (causing the different assertion in the first place) seems incorrect to me.

? conhost.exe is the cmd.exe shell, it is started by `jobstart({string})`. I would expect it to be a child of the job.

blueyed commented:
Ok, conhost is ok then..

Seems weird for a clippy-thing to show up as a child process for the job/cmd.exe, no?

For the test it would make sense to assert ">= 4" processes there then for Windows (this is just for waiting that they are spawned).

blueyed committed:
windows: ok(#children >= 4 and #children <= 5)

justinmk commented:
> Seems weird for a clippy-thing to show up as a child process for the job/cmd.exe, no?

Definitely. But we know that it's some random Microsoft popup-like thing, so it's probably not worth investigating...
 
> For the test it would make sense to assert ">= 4" processes there then for Windows (this is just for waiting that they are spawned).

A range of "between 4 and 5" should be enough to avoid false failures, and still keep the test rather strict.

blueyed commented:
Seems to be good: @erw7 please review / squash. 
I think it might be good to have the process tree output somehow still in case of failures, but we can take it as a snippet from this PR here then also maybe still.

erw7 commented:
I want the range to be 3-5(Probably due to different versions of Windows, but in my environment it will be 3. `conhost.exe` cannot be a child process of `cmd.exe`. It is a child process of `csrss.exe`). I think it looks good.

erw7 committed:
windows: ok(#children >= 3 and #chidlen <= 5)

Depending on the version of Windows, conhost.exe may not be included in
the child process.

erw7 committed:
Change to output status on failure
